In [1]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset('titanic')
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [2]:
df.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


### 1) df.dropna(axis = 1, thresh)
- axis = 1이므로 열을 삭제 (디폴트는 행 삭제 : axis = 0)
- thresh에 입력된 숫자 이상 데이터가 존재하지 않으면 그 행을 삭제한다.
    - 예를 들어 thresh = 100 이면 전체 데이터 891개 중에서 결측치가 791개 이상이라서 남은 데이터가 100개보다 더 적으면 그 행을 drop하는 것

In [4]:
#예시: deck를 삭제하고 싶다. -> 300개 이상의 Non-NaN을 보유한 컬럼만 남기기
df2 = df.dropna(axis = 1, thresh = 300)
print(len(df.columns), len(df2.columns))

15 14


deck는 Non-NaN이 203개뿐이다. 그래서 300개 이상 갖고있어야 한다고 하면 이렇게 drop되어 없어짐

In [5]:
df2.head(3)

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   

     who  adult_male  embark_town alive  alone  
0    man        True  Southampton    no  False  
1  woman       False    Cherbourg   yes  False  
2  woman       False  Southampton   yes   True

### 2) 결측치가 특정 "비율" 이상으로 가지고 있어야만 남기기

In [6]:
df.isna().sum()/len(df)

survived       0.000000
pclass         0.000000
sex            0.000000
age            0.198653
sibsp          0.000000
parch          0.000000
fare           0.000000
embarked       0.002245
class          0.000000
who            0.000000
adult_male     0.000000
deck           0.772166
embark_town    0.002245
alive          0.000000
alone          0.000000
dtype: float64

위는 각 열별 결측치의 비율이다. 근데 결측치가 너무 많은 열은 버리고 싶다.  
예를 들어 결측치 비율이 15%를 넘기면 그 열 자체를 drop하고 싶을 때는..

In [7]:
[df.isna().sum()/len(df) > 0.15]

[survived       False
 pclass         False
 sex            False
 age             True
 sibsp          False
 parch          False
 fare           False
 embarked       False
 class          False
 who            False
 adult_male     False
 deck            True
 embark_town    False
 alive          False
 alone          False
 dtype: bool]

In [8]:
# 15% 넘는 결측치만 갖고 있는 열들 
df.columns[df.isna().sum()/len(df) > 0.15]

Index(['age', 'deck'], dtype='object')

In [9]:
df.drop(columns = df.columns[df.isna().sum()/len(df) > 0.15], axis = 1)

survived  pclass     sex  sibsp  parch     fare embarked   class    who  \
0           0       3    male      1      0   7.2500        S   Third    man   
1           1       1  female      1      0  71.2833        C   First  woman   
2           1       3  female      0      0   7.9250        S   Third  woman   
3           1       1  female      1      0  53.1000        S   First  woman   
4           0       3    male      0      0   8.0500        S   Third    man   
..        ...     ...     ...    ...    ...      ...      ...     ...    ...   
886         0       2    male      0      0  13.0000        S  Second    man   
887         1       1  female      0      0  30.0000        S   First  woman   
888         0       3  female      1      2  23.4500        S   Third  woman   
889         1       1    male      0      0  30.0000        C   First    man   
890         0       3    male      0      0   7.7500        Q   Third    man   

     adult_male  embark_town alive  alone  
0          True  Southampton    no  False  
1         False    Cherbourg   yes  False  
2         False  Southampton   yes   True  
3         False  Southampton   yes  False  
4          True  Southampton    no   True  
..          ...          ...   ...    ...  
886        True  Southampton    no   True  
887       False  Southampton   yes   True  
888       False  Southampton    no  False  
889        True    Cherbourg   yes   True  
890        True   Queenstown    no   True  

[891 rows x 13 columns]